qcodes_contrib_drivers/drivers/Vaunix/LDA.py

```diff
@@ -0,0 +1,291 @@
+r"""
+This is the QCoDeS driver for Vaunix LDA digital attenuators. It requires the
+DLL that comes with the instrument, ``VNX_atten64.dll`` and/or
+``VNX_atten.dll``, for 64-bit Windows and 32-bit Windows, respectively. If the
+instrument has more than one physical channel, ``InstrumentChannel``s are
+created for each one. If the instrument has only one physical channel, no
+channels are created and the parameters will be assigned to this instrument
+instead. The sweep profiles available in the API are not implemented.
+
+Tested with 64-bit system and
+- LDA-133
+- LDA-802Q
+
+"""
+
+import logging
+from typing import Optional, Dict, Callable, Union, cast
+from functools import partial
+from platform import architecture
+import os
+import sys
+import ctypes
+
+from qcodes import Instrument, InstrumentChannel, Parameter
+from qcodes.utils.validators import Numbers
+
+logger = logging.getLogger(__name__)
+
+
+class Vaunix_LDA(Instrument):
+    def __init__(self, name: str,
+                 serial_number: int,
+                 dll_path: Optional[str] = None,
+                 channel_names: Optional[Dict[int, str]] = None,
+                 test_mode: Optional[bool] = False,
+                 **kwargs):
+        r"""
+        QCoDeS Instrument for Vaunix LDA digital attenuators.
+
+        Args:
+            name: Qcodes name for this instrument
+            serial_number: Serial number of the instrument, used to identify
+                it.
+            dll_path: Look for the LDA DLLs in this directory.
+                By default, the directory this file is in.
+            channel_names: Optionally assign these names to the channels.
+            test_mode: If True, communicates with the API but not with the
+                device. For testing purposes.
+        """
+
+        super().__init__(name=name, **kwargs)
+        self.serial_number = serial_number
+        self.reference = None
+
+        if channel_names is None:
+            channel_names = {}
+
+        self.dll = self._get_dll(dll_path)
+        self.dll.fnLDA_SetTestMode(test_mode)  # Test API without communication
+
+        # Find all Vaunix devices, init the one with matching serial number.
+        num_devices = self.dll.fnLDA_GetNumDevices()
+        device_IDs = ctypes.c_int * num_devices
+        device_refs = device_IDs()
+        self.dll.fnLDA_GetDevInfo(device_refs)
+        devices = {self.dll.fnLDA_GetSerialNumber(ref): ref
+                   for ref in device_refs}
+        self.reference = devices.get(self.serial_number, "not found")
+        if self.reference == "not found":
+            raise ValueError(f"LDA with serial number {self.serial_number}"
+                             f" was not found in the system. Found: {devices}")
+
+        self.dll.fnLDA_InitDevice(self.reference)
+
+        num_channels = self.dll.fnLDA_GetNumChannels(self.reference)
+        if num_channels == 1:
+            # don't add Channel objects, add parameters directly instead
```

Comment:
in order not to repeat this code in LdaChannel, it could be extracted into a function that takes `instrument_for_parameters` and `root_instrument` as arguments.

Reply:
Fixed in 5226eac2.